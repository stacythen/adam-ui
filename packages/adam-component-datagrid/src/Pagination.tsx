import React from 'react';

export interface IPaginationProps {
  pageSize: number;
  totalPages: number;
  currentPage: number;
  rowCount: number;
}

interface IPaginationDispatchProps {
  onFirstClicked: () => void;
  onPreviousClicked: () => void;
  onNextClicked: () => void;
  onLastClicked: () => void;
}

type PaginationProps = IPaginationProps & IPaginationDispatchProps;

const Pagination = (props: PaginationProps): React.ReactElement => {
  const {
    pageSize,
    currentPage,
    totalPages,
    rowCount,
    onFirstClicked,
    onPreviousClicked,
    onNextClicked,
    onLastClicked,
  } = props;

  {
    /* << < 1 2 3 > >>
        dropdown items per page
        1 - 10 of 39 items */
  }
  return (
    <>
      <div>
        <button onClick={onFirstClicked} disabled={currentPage === 1}>
          {'<<'}
        </button>
        <button onClick={onPreviousClicked} disabled={currentPage === 1}>
          {'<'}
        </button>
        123
        <button onClick={onNextClicked} disabled={currentPage === totalPages}>
          {'>'}
        </button>
        <button onClick={onLastClicked} id="btLast" disabled={currentPage === totalPages}>
          {'>>'}
        </button>
      </div>
      <div>
        <span>{rowCount ? (currentPage - 1) * pageSize + 1 : 0}</span>
        <span>-</span>
        <span>{Math.min((currentPage - 1) * pageSize + pageSize, rowCount | 0)}</span>
        <span>of {rowCount} items</span>
      </div>
    </>
  );
};

export default Pagination;
