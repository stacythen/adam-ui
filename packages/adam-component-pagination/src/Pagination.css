.pagination-panel {
  margin: 20px 0;
  opacity: 0.77;
  font-size: 9pt;

  padding: 0 18px;
  display: flex;
  justify-content: space-between;
}

.pagination {
  display: inline-block;
  padding: 0;
  margin: 0;
}

.pagination > li {
  display: inline;
  padding: 6px;
  margin: 0 3px;
  cursor: pointer;
}

.pagination > li.pagination-page.active {
  background-color: #3b82de;
  color: #fff;
}

.pagination > li.disabled {
  cursor: not-allowed;
  opacity: 0.38;
}

.page-size-selector {
  margin-left: 64px;
}

.page-size-selector .page-size-select {
  height: 28px;
  line-height: 28px;
  opacity: 0.77;
  border-radius: 4px;
  padding-right: 2px;
}

.page-size-selector .page-size-text {
  margin-left: 8px;
}

.pagination-summary {
  float: right;
}

.paginator {
  border: none;
  background-color: transparent;
  cursor: pointer;
  outline: 0 !important;
  transition: transform 0.25s ease;
}

.paginator:active {
  transform: scale(1.5);
}

.paginator:disabled {
  cursor: not-allowed;
}

.pagination-panel-left {
  flex-grow: 1;
}

.pagination-panel-right {
  display: flex;
  align-items: center;
}

.refresh {
  transform: scaleX(-1);
  margin-left: 6px;
  cursor: pointer;
}

.refresh:active {
  transform: scale(-1.5, 1.5);
}
