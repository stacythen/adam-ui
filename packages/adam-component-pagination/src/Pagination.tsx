import React from 'react';
import { Paginator } from './paginator';
import { IPaginationInfo } from './types';
import PaginationSizeSelector from './PaginationSizeSelector';
import PaginationSummary from './PaginationSummary';

import './Pagination.css';

export interface IPaginationProps {
  pageSize: number;
  currentPage: number;
  rowCount: number;
}

export interface IPaginationConfig {
  pageRange?: number;
}

interface IPaginationDispatchProps {
  onFirstClicked?: () => void;
  onPreviousClicked?: () => void;
  onNextClicked?: () => void;
  onLastClicked?: () => void;
  onPageChanged?: (pageNumber: number, pageSize: number) => void;
}

const PAGE_RANGE = 10;
type PaginationProps = IPaginationProps & IPaginationDispatchProps & IPaginationConfig;

const Pagination = (props: PaginationProps): React.ReactElement => {
  const {
    pageRange = PAGE_RANGE,
    pageSize,
    currentPage,
    rowCount,
    onFirstClicked,
    onPreviousClicked,
    onNextClicked,
    onLastClicked,
    onPageChanged,
  } = props;

  const paginationInfo = Paginator(pageSize, pageRange, rowCount, currentPage);

  const pageChanged = (currentPage: number, pageSize: number): void => {
    if (onPageChanged) {
      onPageChanged(currentPage, pageSize);
    }
  };

  const handlePageSizeChanged = (evt: React.ChangeEvent<HTMLSelectElement>): void => {
    const { currentPage, rowCount } = props;

    /**
     * scenario: totalRecords = 100, pageSize = 10, currentPage = 10 (last page), totalPages = 10
     * When user change pageSize = 20, the totalPages = 5, user will be landed in currentPage = 5
     */
    const newPageSize = parseInt(evt.target.value, 10);
    let newCurrentPage = currentPage;

    const totalPages = Math.ceil(rowCount / newPageSize);
    if (totalPages < newCurrentPage) {
      newCurrentPage = totalPages;
    }

    pageChanged(newCurrentPage, newPageSize);
  };

  const onPaginatorClicked = (newPageNumber: number, callback?: () => void) => () => {
    pageChanged(newPageNumber, pageSize);
    if (callback) {
      callback();
    }
  };

  const buildPages = (paginationInfo: IPaginationInfo) => {
    const { currentPage } = props;

    const pager = (
      <React.Fragment>
        {new Array(paginationInfo.last_page).fill({}).map((_, index) => {
          const i = paginationInfo.first_page + index;
          const isActive = currentPage === i ? ' active' : '';
          return (
            <li key={i.toString()} className={'pagination-page' + isActive} onClick={onPaginatorClicked(i)}>
              {i}
            </li>
          );
        })}
      </React.Fragment>
    );
    return (
      <>
        <button
          name="firstPage"
          className="paginator"
          onClick={onPaginatorClicked(1, onFirstClicked)}
          disabled={!paginationInfo.has_previous_page}
        >
          ≪
        </button>
        <button
          name="prevPage"
          className="paginator"
          onClick={onPaginatorClicked(paginationInfo.previous_page, onPreviousClicked)}
          disabled={!paginationInfo.has_previous_page}
        >
          &lt;
        </button>
        <ul className="pagination">{pager}</ul>
        <button
          name="nextPage"
          className="paginator"
          onClick={onPaginatorClicked(paginationInfo.next_page, onNextClicked)}
          disabled={!paginationInfo.has_next_page}
        >
          &gt;
        </button>
        <button
          name="lastPage"
          className="paginator"
          onClick={onPaginatorClicked(paginationInfo.total_pages, onLastClicked)}
          disabled={!(paginationInfo.has_next_page && paginationInfo.current_page !== paginationInfo.total_pages)}
        >
          ≫
        </button>
      </>
    );
  };

  return (
    <div className={`pagination-panel`} style={{ padding: '0 18px' }}>
      {buildPages(paginationInfo)}
      <PaginationSizeSelector pageSize={pageSize} onPageSizeChanged={handlePageSizeChanged} />
      <PaginationSummary {...paginationInfo} />
    </div>
  );
};

export default Pagination;
