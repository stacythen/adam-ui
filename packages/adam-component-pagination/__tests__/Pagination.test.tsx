import React from 'react';
import { shallow } from 'enzyme';
import Pagination from '../src/Pagination';

describe('Pagination', () => {
  it('should render properly', () => {
    const link = shallow(<Pagination pageSize={10} currentPage={1} rowCount={1000} />);
    expect(link).toMatchSnapshot();
  });

  it('should render properly for page 10', () => {
    const link = shallow(<Pagination pageSize={10} currentPage={10} rowCount={1000} />);
    expect(link).toMatchSnapshot();
  });

  it('should render without page size', () => {
    const link = shallow(<Pagination pageSize={10} currentPage={10} rowCount={1000} hidePageSize={true} />);
    expect(link).toMatchSnapshot();
  });

  it('should render without page summary', () => {
    const link = shallow(<Pagination pageSize={10} currentPage={10} rowCount={1000} hidePageSummary={true} />);
    expect(link).toMatchSnapshot();
  });

  it('should render without refresh button', () => {
    const link = shallow(<Pagination pageSize={10} currentPage={10} rowCount={1000} hideRefresh={true} />);
    expect(link).toMatchSnapshot();
  });
});
