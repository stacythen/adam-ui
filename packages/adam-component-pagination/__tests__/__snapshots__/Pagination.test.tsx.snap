// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination should render properly 1`] = `
<div
  className="pagination-panel"
>
  <div
    className="pagination-panel-left"
  >
    <button
      className="paginator"
      disabled={true}
      name="firstPage"
      onClick={[Function]}
    >
      ≪
    </button>
    <button
      className="paginator"
      disabled={true}
      name="prevPage"
      onClick={[Function]}
    >
      &lt;
    </button>
    <ul
      className="pagination"
    >
      <li
        className="pagination-page active"
        key="1"
        onClick={[Function]}
      >
        1
      </li>
      <li
        className="pagination-page"
        key="2"
        onClick={[Function]}
      >
        2
      </li>
      <li
        className="pagination-page"
        key="3"
        onClick={[Function]}
      >
        3
      </li>
      <li
        className="pagination-page"
        key="4"
        onClick={[Function]}
      >
        4
      </li>
      <li
        className="pagination-page"
        key="5"
        onClick={[Function]}
      >
        5
      </li>
      <li
        className="pagination-page"
        key="6"
        onClick={[Function]}
      >
        6
      </li>
      <li
        className="pagination-page"
        key="7"
        onClick={[Function]}
      >
        7
      </li>
      <li
        className="pagination-page"
        key="8"
        onClick={[Function]}
      >
        8
      </li>
      <li
        className="pagination-page"
        key="9"
        onClick={[Function]}
      >
        9
      </li>
      <li
        className="pagination-page"
        key="10"
        onClick={[Function]}
      >
        10
      </li>
    </ul>
    <button
      className="paginator"
      disabled={false}
      name="nextPage"
      onClick={[Function]}
    >
      &gt;
    </button>
    <button
      className="paginator"
      disabled={false}
      name="lastPage"
      onClick={[Function]}
    >
      ≫
    </button>
    <PaginationSizeSelector
      onPageSizeChanged={[Function]}
      pageSize={10}
    />
  </div>
  <div
    className="pagination-panel-right"
  >
    <div>
      <PaginationSummary
        current_page={1}
        first_page={1}
        first_result={0}
        has_next_page={true}
        has_previous_page={false}
        last_page={10}
        last_result={9}
        next_page={2}
        pages={10}
        previous_page={0}
        results={10}
        total_pages={100}
        total_results={1000}
      />
    </div>
    <div
      className="refresh"
      onClick={[Function]}
    >
      ↺
    </div>
  </div>
</div>
`;

exports[`Pagination should render properly for page 10 1`] = `
<div
  className="pagination-panel"
>
  <div
    className="pagination-panel-left"
  >
    <button
      className="paginator"
      disabled={false}
      name="firstPage"
      onClick={[Function]}
    >
      ≪
    </button>
    <button
      className="paginator"
      disabled={false}
      name="prevPage"
      onClick={[Function]}
    >
      &lt;
    </button>
    <ul
      className="pagination"
    >
      <li
        className="pagination-page"
        key="5"
        onClick={[Function]}
      >
        5
      </li>
      <li
        className="pagination-page"
        key="6"
        onClick={[Function]}
      >
        6
      </li>
      <li
        className="pagination-page"
        key="7"
        onClick={[Function]}
      >
        7
      </li>
      <li
        className="pagination-page"
        key="8"
        onClick={[Function]}
      >
        8
      </li>
      <li
        className="pagination-page"
        key="9"
        onClick={[Function]}
      >
        9
      </li>
      <li
        className="pagination-page active"
        key="10"
        onClick={[Function]}
      >
        10
      </li>
      <li
        className="pagination-page"
        key="11"
        onClick={[Function]}
      >
        11
      </li>
      <li
        className="pagination-page"
        key="12"
        onClick={[Function]}
      >
        12
      </li>
      <li
        className="pagination-page"
        key="13"
        onClick={[Function]}
      >
        13
      </li>
      <li
        className="pagination-page"
        key="14"
        onClick={[Function]}
      >
        14
      </li>
    </ul>
    <button
      className="paginator"
      disabled={false}
      name="nextPage"
      onClick={[Function]}
    >
      &gt;
    </button>
    <button
      className="paginator"
      disabled={false}
      name="lastPage"
      onClick={[Function]}
    >
      ≫
    </button>
    <PaginationSizeSelector
      onPageSizeChanged={[Function]}
      pageSize={10}
    />
  </div>
  <div
    className="pagination-panel-right"
  >
    <div>
      <PaginationSummary
        current_page={10}
        first_page={5}
        first_result={90}
        has_next_page={true}
        has_previous_page={true}
        last_page={14}
        last_result={99}
        next_page={11}
        pages={10}
        previous_page={9}
        results={10}
        total_pages={100}
        total_results={1000}
      />
    </div>
    <div
      className="refresh"
      onClick={[Function]}
    >
      ↺
    </div>
  </div>
</div>
`;

exports[`Pagination should render without page size 1`] = `
<div
  className="pagination-panel"
>
  <div
    className="pagination-panel-left"
  >
    <button
      className="paginator"
      disabled={false}
      name="firstPage"
      onClick={[Function]}
    >
      ≪
    </button>
    <button
      className="paginator"
      disabled={false}
      name="prevPage"
      onClick={[Function]}
    >
      &lt;
    </button>
    <ul
      className="pagination"
    >
      <li
        className="pagination-page"
        key="5"
        onClick={[Function]}
      >
        5
      </li>
      <li
        className="pagination-page"
        key="6"
        onClick={[Function]}
      >
        6
      </li>
      <li
        className="pagination-page"
        key="7"
        onClick={[Function]}
      >
        7
      </li>
      <li
        className="pagination-page"
        key="8"
        onClick={[Function]}
      >
        8
      </li>
      <li
        className="pagination-page"
        key="9"
        onClick={[Function]}
      >
        9
      </li>
      <li
        className="pagination-page active"
        key="10"
        onClick={[Function]}
      >
        10
      </li>
      <li
        className="pagination-page"
        key="11"
        onClick={[Function]}
      >
        11
      </li>
      <li
        className="pagination-page"
        key="12"
        onClick={[Function]}
      >
        12
      </li>
      <li
        className="pagination-page"
        key="13"
        onClick={[Function]}
      >
        13
      </li>
      <li
        className="pagination-page"
        key="14"
        onClick={[Function]}
      >
        14
      </li>
    </ul>
    <button
      className="paginator"
      disabled={false}
      name="nextPage"
      onClick={[Function]}
    >
      &gt;
    </button>
    <button
      className="paginator"
      disabled={false}
      name="lastPage"
      onClick={[Function]}
    >
      ≫
    </button>
  </div>
  <div
    className="pagination-panel-right"
  >
    <div>
      <PaginationSummary
        current_page={10}
        first_page={5}
        first_result={90}
        has_next_page={true}
        has_previous_page={true}
        last_page={14}
        last_result={99}
        next_page={11}
        pages={10}
        previous_page={9}
        results={10}
        total_pages={100}
        total_results={1000}
      />
    </div>
    <div
      className="refresh"
      onClick={[Function]}
    >
      ↺
    </div>
  </div>
</div>
`;

exports[`Pagination should render without page summary 1`] = `
<div
  className="pagination-panel"
>
  <div
    className="pagination-panel-left"
  >
    <button
      className="paginator"
      disabled={false}
      name="firstPage"
      onClick={[Function]}
    >
      ≪
    </button>
    <button
      className="paginator"
      disabled={false}
      name="prevPage"
      onClick={[Function]}
    >
      &lt;
    </button>
    <ul
      className="pagination"
    >
      <li
        className="pagination-page"
        key="5"
        onClick={[Function]}
      >
        5
      </li>
      <li
        className="pagination-page"
        key="6"
        onClick={[Function]}
      >
        6
      </li>
      <li
        className="pagination-page"
        key="7"
        onClick={[Function]}
      >
        7
      </li>
      <li
        className="pagination-page"
        key="8"
        onClick={[Function]}
      >
        8
      </li>
      <li
        className="pagination-page"
        key="9"
        onClick={[Function]}
      >
        9
      </li>
      <li
        className="pagination-page active"
        key="10"
        onClick={[Function]}
      >
        10
      </li>
      <li
        className="pagination-page"
        key="11"
        onClick={[Function]}
      >
        11
      </li>
      <li
        className="pagination-page"
        key="12"
        onClick={[Function]}
      >
        12
      </li>
      <li
        className="pagination-page"
        key="13"
        onClick={[Function]}
      >
        13
      </li>
      <li
        className="pagination-page"
        key="14"
        onClick={[Function]}
      >
        14
      </li>
    </ul>
    <button
      className="paginator"
      disabled={false}
      name="nextPage"
      onClick={[Function]}
    >
      &gt;
    </button>
    <button
      className="paginator"
      disabled={false}
      name="lastPage"
      onClick={[Function]}
    >
      ≫
    </button>
    <PaginationSizeSelector
      onPageSizeChanged={[Function]}
      pageSize={10}
    />
  </div>
  <div
    className="pagination-panel-right"
  >
    <div
      className="refresh"
      onClick={[Function]}
    >
      ↺
    </div>
  </div>
</div>
`;

exports[`Pagination should render without refresh button 1`] = `
<div
  className="pagination-panel"
>
  <div
    className="pagination-panel-left"
  >
    <button
      className="paginator"
      disabled={false}
      name="firstPage"
      onClick={[Function]}
    >
      ≪
    </button>
    <button
      className="paginator"
      disabled={false}
      name="prevPage"
      onClick={[Function]}
    >
      &lt;
    </button>
    <ul
      className="pagination"
    >
      <li
        className="pagination-page"
        key="5"
        onClick={[Function]}
      >
        5
      </li>
      <li
        className="pagination-page"
        key="6"
        onClick={[Function]}
      >
        6
      </li>
      <li
        className="pagination-page"
        key="7"
        onClick={[Function]}
      >
        7
      </li>
      <li
        className="pagination-page"
        key="8"
        onClick={[Function]}
      >
        8
      </li>
      <li
        className="pagination-page"
        key="9"
        onClick={[Function]}
      >
        9
      </li>
      <li
        className="pagination-page active"
        key="10"
        onClick={[Function]}
      >
        10
      </li>
      <li
        className="pagination-page"
        key="11"
        onClick={[Function]}
      >
        11
      </li>
      <li
        className="pagination-page"
        key="12"
        onClick={[Function]}
      >
        12
      </li>
      <li
        className="pagination-page"
        key="13"
        onClick={[Function]}
      >
        13
      </li>
      <li
        className="pagination-page"
        key="14"
        onClick={[Function]}
      >
        14
      </li>
    </ul>
    <button
      className="paginator"
      disabled={false}
      name="nextPage"
      onClick={[Function]}
    >
      &gt;
    </button>
    <button
      className="paginator"
      disabled={false}
      name="lastPage"
      onClick={[Function]}
    >
      ≫
    </button>
    <PaginationSizeSelector
      onPageSizeChanged={[Function]}
      pageSize={10}
    />
  </div>
  <div
    className="pagination-panel-right"
  >
    <div>
      <PaginationSummary
        current_page={10}
        first_page={5}
        first_result={90}
        has_next_page={true}
        has_previous_page={true}
        last_page={14}
        last_result={99}
        next_page={11}
        pages={10}
        previous_page={9}
        results={10}
        total_pages={100}
        total_results={1000}
      />
    </div>
  </div>
</div>
`;
